types/react/index.d.ts

```diff
@@ -264,6 +265,24 @@ declare namespace React {
         props?: Q, // should be Q & Attributes
         ...children: ReactNode[]): ReactElement<P>;
 
+    // Context via RenderProps
+    type Provider<T> = ComponentType<{
+        value: T,
+        children?: ReactNode
+    }>;
+    type Consumer<T> = ComponentType<{
+        children: (value: T) => ReactNode,
+        unstable_observedBits?: number
+    }>;
+    interface Context<T> {
+        Provider: Provider<T>;
+        Consumer: Consumer<T>;
+    }
+    function createContext<T>(
+        defaultValue: T,
+        calculateChangedBits?: (prev: T, next: T) => number
+    ): Context<T>;
```

Comment:
It should be possible to call `createContext()` to create one with `defaultValue` of `undefined`

`function createContext<T = undefined>(defaultValue: T, ...)`?

Reply:
Hmm. I'm confused here.  Calling `createContext()` (*i.e.,* no arguments) would require the `defaultValue` to be optional.  Otherwise you'd need to do an explicit `createContext(undefined)`.  Also, since `T` can be anything, I don't understand `T|undefined`.

FYI, I address this case of optional default values in [my article on the new Context API](https://medium.com/@mtiller/react-16-3-context-api-intypescript-45c9eeb7a384) (using the typings as they stand here, but wrapping them in one additional layer).  Also note that, in my opinion (see article), it isn't *just* about making the default value optional.  But again, I used these typings without issue.

Reply:
OK, rebased this post merge conflict.

Spent some time with this tonight and it looks like the behavior of the final API is that 
`createContext()`  gives you a context object where `_defaultValue` is undefined. 

`createContext('foo')` gives you a context object where `_defaultValue` is `foo`

Given that, I'm a bit confused as to the correct typing for `createContext()` with no params. I tried the overload @Kovensky mentioned but doing so types the resulting RenderProps function args with `{} or undefined`, which I don't think is correct. 

I suspect the overload closest to the API's usage would be

```
function createContext(): Context<any>
```

Which completely gives up type safety for RenderProps function but at least allows it to be used. `{} or undefined` doesn't allow you to still give, say, an anonymous object in the Provider's `value` field and still be able to retrieve it later in the RenderProps.

If anyone has any opinions on this they'd be welcome.

Reply:
It resulting in `RenderProps` with `{}` was completely intended. You should call `createContext<MyContextType>()`, not just `createContext()`; similar to how you should call `new Set<valueType>()` instead of `new Set()`.

Reply:
Although there is an argument for making it possible to `createContext<MyContextType>()` produce a function with render props of `MyContextType` instead of `MyContextType|undefined`; in this case, your provider's `value` type would have to be required and not include `undefined`.

Reply:
@Kovensky Yes, I agree. Thanks for the explanation. I've made it `T | undefined` for now.

Reply:
Please correct me if I am wrong, but having the resulting `T` as `{}|undefined` means that to check for `undefined` in every Consumer use, right? I haven't actually tested that yet.

```tsx
const context = createContext<MyContextType>()

<Provider value={myValue}>
  <Consumer>{value => if (value) { ... }}</Consumer>
</Provider>
```

If this is the case, then it's definitely a bad way to go imo. Sadly, my TypeScript knowledge is limited and I am not sure about better way here.

Reply:
@FredyC this is the intended behavior since React doesn't enforce that a `Consumer` be a descendant of its dual `Provider` so if you do not provide a default value on creating the context, there is a chance that it is `undefined`, since the React tree is determined at runtime there's no way to statically check that

@scally with TS 2.8 we can now use the overload while preserving the required value type of `Provider` using a conditional type, that way we can still have no-parameter defaultValue that the user can manually assign the generic type and `Provider.value` is still required, see cod below:
```tsx
import { ComponentType, ReactNode } from 'react'

type Consumer<T> = ComponentType<{
    children(value: T): ReactNode
}>
type Provider<T> = ComponentType<{
    value: T extends undefined ? never : T
    children?: ReactNode
}>

type Context<T> = {
    Consumer: Consumer<T>
    Provider: Provider<T>
}

declare function createContext<T>(defaultValue: T): Context<T>
declare function createContext<T>(): Context<T | undefined>

const Foo = createContext<string>()
const Bar = createContext('hello')

const foo = (
    <Foo.Provider value={undefined /* error */}>
        <Foo.Consumer>{val => val /* val: T | undefined */}</Foo.Consumer>
    </Foo.Provider>
)

const bar = (
    <Bar.Provider value="hello">
        <Bar.Consumer>{val => val /* T */}</Bar.Consumer>
    </Bar.Provider>
)
```

Reply:
I am failing to see how this can actually work (from your first example). Obviously second overload is used so `T = MyContextType | undefined`. If you supply value later with `Provider`, the T is not gonna change, unless there is some magic part I am missing.

```tsx
const Context = createContext<MyContextType>()

<Context.Provider value={myValue>
  <Context.Consumer>{value => { 
    /* value here will be defined, even though it's typed as optional */ }}
  </Context.Consumer>
</Context.Provider>
```

Either way, I am more concerned about second use case when you don't actually use `Provider` intentionally and supply a value to `createContext` which I understand is a completely legal pattern. In that case developer is "punished" for this and forced to check for `undefined` everytime while he is certain, that value is there.

---

I am curious, would this work the same way? It could possibly serve as an escape hatch for the scenarios above. Only caveat is that I would need to specify the same type twice, but it's at a single spot in a code so it's not that big deal.

```ts
function createContext<T, S extends T | undefined>(): Context<S>;
```

Reply:
Yup, I'm just saying that we shouldn't want to correct that behavior, because you can always use `Consumer` not paired with a `Provider` so the value of `T` should not change even if you later pass in the value as a prop to `Provider`. 

I.e. if you create the context with the default value, you are *always* guaranteed to have a defined value wherever you use `Consumer` (even not as a descendant of `Provider`), but once you create the context without a default value, you are not guaranteed that defined value unless you wrap the `Consumer` inside a `Provider`, which wouldn't be able to be statically analyzed before runtime.

I'm agreeing with you that it's unfortunate that we'll always have to check for undefined, but I think that should be the way the interface is structured due to the lack of compile-time guarantees of the consumer value

Reply:
Well, in your first example you had the comment there `/* value here will be defined, even though it's typed as optional */` and that was what's confused me the most. As you have said now, you will have to check for `undefined` either way. I understand now.

I see a bit of problem arising here. Developers are generally lazy and we might end up with pattern like this.

```tsx
interface MyContextType { myValue: string }
const Context = createContext<MyContextType>()

<Context.Consumer>{value => { 
  <div>{value!.myValue.toUpperCase()}</div> // <-- notice the "!" mark there
</Context.Consumer>
```

That might be a good source of runtime errors (_and blaming TypeScript/React for it_) in case someone really forgets to add the Provider. It's true that if anyone is using `!` it's for own risk, but there are people that might just copy&paste and then be surprised.

---

What about this approach? Would this work the same way? I am not sure if it's a correct syntax, but you get the gist. This would at least allow to override this `undefined` type in case someone knows what they are doing.

```ts
function createContext<T, S = T | undefined>(): Context<S>;
```

Reply:
Hmm.. it might not be immediately clear to the user why the second generic type param may be used, and they'll have to dig into the type defs to actually find out how they resolve.

With my suggestion (the two overloads), creating context with a default param will assert the value in the consumer to be always be defined:
```tsx
interface ContextProps {
  foo: string
  bar?: number
}

// note the default values passed in as the first parameter
const Context = createContext<Props>({ foo: 'foo' }) // typed as Props
const OtherContext = createContext({ foo: 'foo' }) // inferred to be type { foo: string }

// even without a Provider it is still typed to be required
const app = (
  <App>
    <Context.Consumer>
      {( value ) => /* value is typed to be Props here, no need to check for undefined */ }
    </Context.Consumer>
    <Context.Consumer>
      {( value ) => /* value is typed to be { foo: string } here, no need to check for undefined */ }
    </Context.Consumer>
  </App>
)
```

in this case, all you need as a type consumer to override the `undefined` behavior is to just provide a default value, the type inference will work for you or you can explicitly set the generic type parameter, and as long as the default value is passed in as a parameter, the consumer render props will always be defined

Reply:
@ferdaber I'm trying to use the TypeScript 2.8 features you mentioned in this definition, but it looks like that's not possible yet? 

I've updated the headers to `// TypeScript Version: 2.8` but that results in a linting error.

According to this comment, we can only use TypeScript up to 2.7 right now, so I may not be able to use the definition you specify.

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24586#issuecomment-376911352

Reply:
ah, that's too bad, we'll keep it in the backburner then